Add stricter type checking to the Web Extension code generator.

xeenon opened the pull request:
#### bc22aaaf81abef67e13f40d895cf7e51653be094
<pre>
Add stricter type checking to the Web Extension code generator.
<a href="https://webkit.org/b/261386">https://webkit.org/b/261386</a>
rdar://problem/115255883

Reviewed by Brian Weinstein.

* Have the code generator do strict type checks per argument now that the argument handling
  is simplified and easier to add those checks. This aligns with Chrome and Firefox type checks.
* Simplified validateDictionary to not take optionalKeys, this reduces the number of static
  NSSets we have for each API. Now only the dictionary of types is required for most APIs.
* Tweak the wording to not say &quot;string value&quot; and &quot;number value&quot;, to be just &quot;string&quot; and &quot;number&quot;.
* Have _WKWebExtensionWebNavigationURLFilter use the new toErrorString and validateObject functions,
  which was missed in my earlier error message refactoring.

* Source/WebKit/Shared/Extensions/WebExtensionUtilities.h:
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.mm:
(WebKit::classToClassString): Changed &quot;string value&quot; and &quot;number value&quot;, to be just &quot;string&quot; and &quot;number&quot;.
(WebKit::validateDictionary): Dropped optionalKeys, and generate it from the types and required set.
(WebKit::validateObject): Added an ASSERT.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIAlarmsCocoa.mm:
(WebKit::WebExtensionAPIAlarms::createAlarm): Removed optionalKeys.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm:
(WebKit::WebExtensionAPIPermissions::parseDetailsDictionary): Removed optionalKeys.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITabsCocoa.mm:
(WebKit::WebExtensionAPITabs::parseTabCreateOptions): Removed optionalKeys.
(WebKit::WebExtensionAPITabs::parseTabUpdateOptions): Removed optionalKeys.
(WebKit::WebExtensionAPITabs::parseTabQueryOptions): Removed optionalKeys.
(WebKit::WebExtensionAPITabs::reload): Removed optionalKeys.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWindowsCocoa.mm:
(WebKit::WebExtensionAPIWindows::parsePopulateTabs): Removed optionalKeys.
(WebKit::WebExtensionAPIWindows::parseWindowTypesFilter): Removed optionalKeys.
(WebKit::WebExtensionAPIWindows::parseWindowCreateOptions): Removed optionalKeys.
(WebKit::WebExtensionAPIWindows::parseWindowUpdateOptions): Removed optionalKeys.
* Source/WebKit/WebProcess/Extensions/Bindings/Scripts/CodeGeneratorExtensions.pm:
(_generateImplementationFile): Added calls to _installArgumentTypeExceptions for arguments.
(_installArgumentTypeExceptions): Added.
(_javaScriptTypeCondition): Only allow null and undefined when the argument is optional. This matches Chrome and Firefox.
(_platformTypeVariableDeclaration): Drop condition for numbers since _installArgumentTypeExceptions handles it now.
* Source/WebKit/WebProcess/Extensions/Cocoa/_WKWebExtensionWebNavigationURLFilter.mm:
(-[_WKWebExtensionWebNavigationURLPredicate initWithTypeString:value:outErrorMessage:]): Use validateObject and toErrorString.
(-[_WKWebExtensionWebNavigationURLFilter initWithDictionary:outErrorMessage:]): Drop urlOptionalKeys.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIAlarms.mm:
(TestWebKitAPI::TEST): Updated error string expectations.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIPermissions.mm:
(TestWebKitAPI::TEST): Updated error string expectations.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITabs.mm:
(TestWebKitAPI::TEST): Updated error string expectations. Added back some error checks that work again with generator changes.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIWebNavigation.mm:
(TestWebKitAPI::TEST): Updated error string expectations.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIWindows.mm:
(TestWebKitAPI::TEST): Updated error string expectations.

Canonical link: <a href="https://commits.webkit.org/267883@main">https://commits.webkit.org/267883@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/088096ef67e3810058c0e0b9beb01abb5d1ba14b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17963 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/18289 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/18854 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/19793 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/16810 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/18159 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/21584 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/18446 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/19793 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/18179 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/18436 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/15627 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/20665 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/15663 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/16380 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/22908 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/16679 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/16549 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20776 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/17112 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/14501 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/16212 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/20572 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2204 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/16960 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

b-weinstein commented on Source/WebKit/Shared/Extensions/WebExtensionUtilities.mm:
```diff
@@ -223,11 +223,15 @@ static bool validate(NSString *key, NSObject *value, id expectedValueType, NSStr
-    NSSet<NSString *> *optionalKeysSet = [[NSSet alloc] initWithArray:optionalKeys];
+    NSMutableSet<NSString *> *optionalKeysSet = [[NSMutableSet alloc] initWithArray:keyTypes.allKeys];
+
+    [optionalKeysSet minusSet:requiredKeysSet];
```
Nit: I'd put these two lines together. Maybe a blank line before creating `optionalKeySet`? I just think it makes it more clear both are needed for the proper set up of that variable.

b-weinstein commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIAlarmsCocoa.mm:
```diff
@@ -85,19 +85,13 @@
     };
 
-    if (!validateDictionary(alarmInfo, @"info", nil, optionalKeys, types, outExceptionString))
+    if (!validateDictionary(alarmInfo, @"info", nil, types, outExceptionString))
```
Are there no required keys for this method? What happens if you pass an empty object?

xeenon replied:
Nothing required, it defaults to a 1 minute alarm.

xeenon committed:
Add stricter type checking to the Web Extension code generator.
https://webkit.org/b/261386
rdar://problem/115255883

Reviewed by Brian Weinstein.

* Have the code generator do strict type checks per argument now that the argument handling
  is simplified and easier to add those checks. This aligns with Chrome and Firefox type checks.
* Simplified validateDictionary to not take optionalKeys, this reduces the number of static
  NSSets we have for each API. Now only the dictionary of types is required for most APIs.
* Tweak the wording to not say "string value" and "number value", to be just "string" and "number".
* Have _WKWebExtensionWebNavigationURLFilter use the new toErrorString and validateObject functions,
  which was missed in my earlier error message refactoring.

* Source/WebKit/Shared/Extensions/WebExtensionUtilities.h:
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.mm:
(WebKit::classToClassString): Changed "string value" and "number value", to be just "string" and "number".
(WebKit::validateDictionary): Dropped optionalKeys, and generate it from the types and required set.
(WebKit::validateObject): Added an ASSERT.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIAlarmsCocoa.mm:
(WebKit::WebExtensionAPIAlarms::createAlarm): Removed optionalKeys.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm:
(WebKit::WebExtensionAPIPermissions::parseDetailsDictionary): Removed optionalKeys.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITabsCocoa.mm:
(WebKit::WebExtensionAPITabs::parseTabCreateOptions): Removed optionalKeys.
(WebKit::WebExtensionAPITabs::parseTabUpdateOptions): Removed optionalKeys.
(WebKit::WebExtensionAPITabs::parseTabQueryOptions): Removed optionalKeys.
(WebKit::WebExtensionAPITabs::reload): Removed optionalKeys.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWindowsCocoa.mm:
(WebKit::WebExtensionAPIWindows::parsePopulateTabs): Removed optionalKeys.
(WebKit::WebExtensionAPIWindows::parseWindowTypesFilter): Removed optionalKeys.
(WebKit::WebExtensionAPIWindows::parseWindowCreateOptions): Removed optionalKeys.
(WebKit::WebExtensionAPIWindows::parseWindowUpdateOptions): Removed optionalKeys.
* Source/WebKit/WebProcess/Extensions/Bindings/Scripts/CodeGeneratorExtensions.pm:
(_generateImplementationFile): Added calls to _installArgumentTypeExceptions for arguments.
(_installArgumentTypeExceptions): Added.
(_javaScriptTypeCondition): Only allow null and undefined when the argument is optional. This matches Chrome and Firefox.
(_platformTypeVariableDeclaration): Drop condition for numbers since _installArgumentTypeExceptions handles it now.
* Source/WebKit/WebProcess/Extensions/Cocoa/_WKWebExtensionWebNavigationURLFilter.mm:
(-[_WKWebExtensionWebNavigationURLPredicate initWithTypeString:value:outErrorMessage:]): Use validateObject and toErrorString.
(-[_WKWebExtensionWebNavigationURLFilter initWithDictionary:outErrorMessage:]): Drop urlOptionalKeys.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIAlarms.mm:
(TestWebKitAPI::TEST): Updated error string expectations.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIPermissions.mm:
(TestWebKitAPI::TEST): Updated error string expectations.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITabs.mm:
(TestWebKitAPI::TEST): Updated error string expectations. Added back some error checks that work again with generator changes.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIWebNavigation.mm:
(TestWebKitAPI::TEST): Updated error string expectations.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIWindows.mm:
(TestWebKitAPI::TEST): Updated error string expectations.

Canonical link: https://commits.webkit.org/267883@main
